### PR TITLE
[Feature:TAGrading] Added Stats for Verify grader

### LINKED ISSUE #2440: Stats for Verify Grader
Add to Stats page for each electronic gradeable:

at the top of the left side (with overall % graded):

% of all components that have been verified (a.k.a. double graded - grader2 != NULL)

% of all components that grader1 is a full access or instructor OR grader 2 != NULL and grader2 is a full access or instructor.

-----------

On the left side, in the "by XX section" list.
instead of "Section 1: 9/10 (90%)" put "Section 1: 9/10 (90% graded), 5/10 (50% verified), 4/10 (40% double graded).

---------

On the right side,  in the 
"Statistics for Manually Graded Components:" list
instead of "Count: 18", put "18/20 (90% graded), 8/20 (40% verified), 6/20 (30% double graded)"

### PR DESCRIPTION
**Pull request by benlee04**
### Please check if the PR fulfills these requirements:

* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant
* [ ] Screenshots are attached to Github PR if visual/UI changes were made

### What is the current behavior?
Grading statistics for a gradeable do not show the percentage of verified assignments in each section.

### What is the new behavior?
The statistics page will display the percentage of verified assignments for each section.
image

### Other information?

closes #2440

### CODE REVIEW

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 1]
```diff
@@ -3335,6 +3362,7 @@ private function getMarkStats(Mark $mark, User $grader) {
      */
     private function getStats(Gradeable $gradeable, User $grader, bool $full_stats, &$total_graded, &$total_total) {
         $num_components = $this->core->getQueries()->getTotalComponentCount($gradeable->getId());
+        //$verified_components = $this->core->getQueries()->getTotalVerifiedComponentCount($gradeable->getId());
```
Comment (Youssef-Chip) [inline site/app/controllers/grading/ElectronicGraderController.php:3365]: can you remove this line and every similar unnecessary comment in this PR. Also can you make sure your package isn't different then the one from main

**Review by Youssef-Chip — CHANGES_REQUESTED** [review]

File: site/app/libraries/database/DatabaseQueries.php [hunk 2]
```diff
@@ -2147,6 +2147,17 @@ public function getTotalComponentCount($g_id) {
...
+    /**
+     * Get an array of Teams for a Gradeable matching the given rotating sections
+     *
+     * @param  string $g_id
+     * @return int
+     */
+    public function getTotalVerifiedComponentCount($g_id) {
+        $this->course_db->query("SELECT count(*) AS cnt FROM gradeable_component_data WHERE GCD_VERIFIER_ID IS NOT NULL", [$g_id]);
+        return intval($this->course_db->row()['cnt']);
+    }
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2159]: It would be better to use native PHP type annotations here, since these PHPDoc comments don't provide any information a native annotation wouldn't provide.

```suggestion
    /**
     * Get an array of Teams for a Gradeable matching the given rotating sections
     */
    public function getTotalVerifiedComponentCount(string $g_id): int {
        $this->course_db->query("SELECT count(*) AS cnt FROM gradeable_component_data WHERE GCD_VERIFIER_ID IS NOT NULL", [$g_id]);
        return intval($this->course_db->row()['cnt']);
    }
```

File: site/app/libraries/database/DatabaseQueries.php [hunk 3]
```diff
@@ -2147,6 +2147,17 @@ public function getTotalComponentCount($g_id) {
         return intval($this->course_db->row()['cnt']);
     }
 
+    /**
+     * Get an array of Teams for a Gradeable matching the given rotating sections
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2151]: This says it gets an array, but the function actually returns an int...

File: site/app/libraries/database/DatabaseQueries.php [hunk 4]
```diff
@@ -2193,6 +2204,62 @@ public function getGradedComponentsCountByGradingSections($g_id, $sections, $sec
...
+     *
+     * @param  string $g_id
+     * @param  array<int>  $sections
+     * @param  string $section_key
+     * @param  bool   $is_team
+     * @return array<int|string,int>
+     */
+    public function getVerifiedComponentsCountByGradingSections($g_id, $sections, $section_key, $is_team) {
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2217]: Same comment here:
```suggestion
     * @param  array<int>  $sections
     * @return array<int|string,int>
     */
    public function getVerifiedComponentsCountByGradingSections(string $g_id, array $sections, string $section_key, bool $is_team): array {
```

File: site/package-lock.json [hunk 5]
```diff
@@ -34,7 +34,7 @@
         "jquery.are-you-sure": "1.9.0",
         "js-cookie": "3.0.5",
         "jspdf": "^2.5.1",
-        "luxon": "^3.4.3",
+        "luxon": "3.4.3",
```
Comment (williamjallen) [inline site/package-lock.json:37]: This change isn't related to this PR and should be reverted.

File: site/app/libraries/database/DatabaseQueries.php [hunk 6]
```diff
@@ -2193,6 +2204,62 @@ public function getGradedComponentsCountByGradingSections($g_id, $sections, $sec
...
+ SELECT * FROM gradeable_data AS gd
+ INNER JOIN (SELECT g_id, $user_or_team_id, max(active_version) as active_version FROM electronic_gradeable_version GROUP BY g_id, $user_or_team_id) AS egd on egd.g_id = gd.g_id AND egd.{$user_or_team_id} = gd.gd_{$user_or_team_id}
+ LEFT JOIN (
+ gradeable_component_data AS gcd
+ INNER JOIN gradeable_component AS gc ON gc.gc_id = gcd.gc_id AND gc.gc_is_peer = {$this->course_db->convertBoolean(false)}
+ )AS gcd ON gcd.gd_id = gd.gd_id WHERE gcd.g_id=? AND GCD_VERIFIER_ID IS NOT NULL
+) AS gd ON {$u_or_t}.{$user_or_team_id} = gd.gd_{$user_or_team_id}
+{$where}
+GROUP BY {$u_or_t}.{$section_key}
+ORDER BY {$u_or_t}.{$section_key}",
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2247]: It looks like you copied this code from elsewhere in this file.  Copying code is not conducive to future maintainability.  Is there a way for you to share more code with the original source?

**Review by williamjallen — CHANGES_REQUESTED** [review]

**Review by Youssef-Chip — CHANGES_REQUESTED** [review]
Can you also make the verified stats only appear when there is at least one submission verified, else don't show anything because it is a bit bulky ( we've discussed this during the meeting)

File: site/app/libraries/database/DatabaseQueries.php (hunk 6 above)
Comment (benjaminlee755) [reply site/app/libraries/database/DatabaseQueries.php:2241]: I referenced off of another code in the file because I saw that this code would work if I made a slight change to it to get the info I needed. I wasn't too sure how to query the database for the same info

**Review by Youssef-Chip — CHANGES_REQUESTED** [review]
When you get the verified components, they unfortunately don't include the non -bad submission. Can you make sure $non_late_verified_components works?

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 7]
```diff
@@ -757,11 +765,14 @@ public function showStatus($gradeable_id) {
                         }
                         $sections['all']['total_components'] += $value * $num_components;
                         $sections['all']['graded_components'] += isset($graded_components[$key]) ? $graded_components[$key] : 0;
+                        $sections['all']['verified_components'] += isset($verified_components[$key]) ? $verified_components[$key] : 0;
```
Comment (williamjallen) [inline site/app/controllers/grading/ElectronicGraderController.php:768]: This is a great opportunity to use PHP's [null coalescing operator](https://www.php.net/manual/en/language.operators.comparison.php#language.operators.comparison.coalesce)!

```suggestion
                        $sections['all']['verified_components'] += $verified_components[$key]) ?? 0;
```

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 8]
```diff
@@ -793,11 +808,14 @@ public function showStatus($gradeable_id) {
...
```
Comment (williamjallen) [inline site/app/controllers/grading/ElectronicGraderController.php:811]: Same here...

```suggestion
                        $sections['all']['verified_components'] += $verified_components[$key] ?? 0;
```

File: site/app/libraries/database/DatabaseQueries.php [hunk 9]
```diff
@@ -178,6 +178,34 @@ public function getUsersByIds(array $user_id_list): ?array {
...
+     *
+     * @param string $u_or_t
+     * @param string $section_key
+     * @param string $users_or_teams
+     * @param string $user_or_team_id
+     * @param string $verified
+     * @param string $where
+     * @return string
+     */
+    public function getGrading($u_or_t, $section_key, $users_or_teams, $user_or_team_id, $verified, $where): ?string {
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:192]: This should probably also be a private function given that it should never be executed from outside this file.

File: site/app/libraries/database/DatabaseQueries.php [hunk 10]
```diff
@@ -2248,10 +2276,22 @@ public function getTotalComponentCount($g_id) {
         return intval($this->course_db->row()['cnt']);
     }
 
+    /**
+     * Returns the total number of verified components found in the database
+     *
+     * @param  string $g_id
+     * @return int
+     */
+    public function getTotalVerifiedComponentCount(string $g_id): int {
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2285]: There's no need to specify PHPDoc type annotations when you already have native type annotations.
```suggestion
     */
    public function getTotalVerifiedComponentCount(string $g_id): int {
```

File: site/app/libraries/database/DatabaseQueries.php [hunk 11]
```diff
@@ -2248,10 +2276,22 @@ public function getTotalComponentCount($g_id) {
...
+        $this->course_db->query("SELECT count(*) AS cnt FROM gradeable_component_data WHERE GCD_VERIFIER_ID IS NOT NULL", [$g_id]);
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2286]: You are passing a SQL parameter here, but there aren't any placeholders in the SQL query.

**Review by williamjallen — CHANGES_REQUESTED** [review]

**Review by VVVITH — APPROVED** [review]
Tested and fixed the codes, all good. Tested basic Verify grader feature, but not make a detail check on specific numbers like percentage, but codes are all converted from similar feature, the algo for stats should be fine.

File: site/cypress/e2e/Cypress-Gradeable/grading_stats.spec.js
Comment (williamjallen) [inline site/cypress/e2e/Cypress-Gradeable/grading_stats.spec.js:1]: It would be good to add some basic Cypress tests for this feature.

File: site/app/views/grading/ElectronicGraderView.php [hunk 12]
```diff
@@ -362,7 +370,7 @@ public function statusPage(
             "include_null_section" => array_key_exists('include_null_section', $_COOKIE) ? $_COOKIE['include_null_section'] : 'omit',
             "warnings" => $warnings,
             "submissions_in_queue" => $submissions_in_queue,
-            "can_manage_teams" => $this->core->getAccess()->canI('grading.electronic.show_edit_teams', ["gradeable" => $gradeable])
+            "can_manage_teams" => $this->core->getAccess()->canI('grading.electronic.show_edit_teams', ["gradeable" => $gradeable]),
```
Comment (williamjallen) [inline site/app/views/grading/ElectronicGraderView.php:373]: This change appears to be unrelated to this PR and should be reverted.

File: site/app/views/grading/ElectronicGraderView.php
Comment (williamjallen) [inline site/app/views/grading/ElectronicGraderView.php:1]: It looks like there are some potential division-by-zero issues here, but the new changes mimic existing code and do not add any additional potential problems.  It would be good to clean up these potential issues in a separate PR.

File: site/app/libraries/database/DatabaseQueries.php [hunk 13]
```diff
@@ -2449,7 +2498,7 @@ public function getGradedComponentsCountByGradingSections($g_id, $sections, $sec
     public function getBadGradedComponentsCountByGradingSections($g_id, $sections, $section_key, $is_team) {
         //getBadTeamSubmissionsByGradingSection
         //getBadUserSubmissionsByGradingSection
-         $u_or_t = "u";
+        $u_or_t = "u";
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2501]: Although this change is a change in a positive direction, it's unrelated to this PR and should be reverted to reduce the risk of conflicts, as well as make the real changes in this PR clear.

File: site/app/libraries/database/DatabaseQueries.php [hunk 14]
```diff
@@ -2407,21 +2407,23 @@ public function getGradedComponentsCountByGradingSections($g_id, $sections, $sec
...
+            INNER JOIN (SELECT g_id, $user_or_team_id, max(active_version) as active_version FROM electronic_gradeable_version GROUP BY g_id, $user_or_team_id) AS egd on egd.g_id = gd.g_id AND egd.{$user_or_team_id} = gd.gd_{$user_or_team_id}
+            LEFT JOIN (
+            gradeable_component_data AS gcd
+            INNER JOIN gradeable_component AS gc ON gc.gc_id = gcd.gc_id AND gc.gc_is_peer = {$this->course_db->convertBoolean(false)}
+            )AS gcd ON gcd.gd_id = gd.gd_id WHERE gcd.g_id=?
+            ) AS gd ON {$u_or_t}.{$user_or_team_id} = gd.gd_{$user_or_team_id}
+            {$where}
+            GROUP BY {$u_or_t}.{$section_key}
+            ORDER BY {$u_or_t}.{$section_key}
+        ",
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2426]: Same comment here about general formatting changes.  By making changes to every line in this query, the real changes being made are hidden in the middle of a large diff.  Reformatting our SQL is a separate issue, which can be addressed in a separate PR if desired.

File: site/app/libraries/database/DatabaseQueries.php [hunk 15]
```diff
@@ -2437,6 +2439,53 @@ public function getGradedComponentsCountByGradingSections($g_id, $sections, $sec
...
+        $id = "user_id";
+        if ($is_team) {
+            $unit = "gradeable_teams";
+            $id = "team_id";
+        }
+
+        $this->course_db->query(
+            "
+            SELECT 
+                " . $unit . "." . $section_key . ", COUNT(" . $unit . "." . $id . ") as verified_components_count
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2460]: `$section_key` is being passed into this function and directly inserted in the SQL query.  Although it looks like it can only ever be fixed strings right now, it's a security liability.  It would be better to be more defensive here and assert that `$section_key` is one of a set of expected values.

Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2460]: Since this is a double-quoted string, it would be more readable to put the variables in the string directly, instead of cluttering the query with string concatenations everywhere.

File: site/app/libraries/database/DatabaseQueries.php [hunk 16]
```diff
@@ -2437,6 +2439,53 @@ public function getGradedComponentsCountByGradingSections($g_id, $sections, $sec
...
+                gradeable_data as gd
+                JOIN gradeable_component_data as gcd ON (gd.gd_id = gcd.gd_id)
+                JOIN " . $unit . " ON (gd.gd_" . $id . " = " . $unit . "." . $id . ")
+            WHERE
+                gd.g_id = ?
+                AND CAST (" . $unit . "." . $section_key . " AS TEXT) = ANY (ARRAY[" . implode(",", array_map(
+                    function ($v) {
+                        return "'" . $v . "'";
+                    },
+                    $sections
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2471]: `$sections` is user-controlled, and should not be inserted directly into a query.  Use a prepared statement instead.  `$this->createParameterList)` will also be useful to generate a list of question marks.

**Review by williamjallen — CHANGES_REQUESTED** [review]

File: site/app/libraries/database/DatabaseQueries.php (hunk 15 above)
Comment (dorian451) [reply site/app/libraries/database/DatabaseQueries.php:2460]: I kinda didn't realize format strings were a thing in PHP, my bad

File: site/cypress/e2e/Cypress-Gradeable/grading_stats.spec.js
Comment (williamjallen) [reply site/cypress/e2e/Cypress-Gradeable/grading_stats.spec.js:1]: Sounds good.  If there are no verified grades in the sample database, I agree that that we should add some in a separate PR.  Do you want to make an issue for that?

File: site/app/libraries/database/DatabaseQueries.php [hunk 17]
```diff
@@ -2433,21 +2433,23 @@ public function getGradedComponentsCountByGradingSections($g_id, $sections, $sec
...
+LEFT JOIN (
+gradeable_component_data AS gcd
+INNER JOIN gradeable_component AS gc ON gc.gc_id = gcd.gc_id AND gc.gc_is_peer = {$this->course_db->convertBoolean(false)}
+)AS gcd ON gcd.gd_id = gd.gd_id WHERE gcd.g_id=?
 ) AS gd ON {$u_or_t}.{$user_or_team_id} = gd.gd_{$user_or_team_id}
 {$where}
 GROUP BY {$u_or_t}.{$section_key}
-ORDER BY {$u_or_t}.{$section_key}",
+ORDER BY {$u_or_t}.{$section_key}
+        ",
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2452]: Still a lot of unrelated whitespace changes here.  These should be reverted before this PR is merged.

File: site/app/libraries/database/DatabaseQueries.php [hunk 18]
```diff
@@ -2463,6 +2465,52 @@ public function getGradedComponentsCountByGradingSections($g_id, $sections, $sec
...
+            "
+            SELECT 
+                $unit.$section_key, COUNT($unit.$id) as verified_components_count
+            FROM 
+                gradeable_data as gd
+                JOIN gradeable_component_data as gcd ON (gd.gd_id = gcd.gd_id)
+                JOIN $unit ON (gd.gd_$id = $unit.$id)
+            WHERE
+                gd.g_id = ?
+                AND CAST ($unit.$section_key AS TEXT) = ANY (ARRAY[" . implode(',', array_fill(0, count($sections), '?')) . "])
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2497]: A few comments about this line:
1. The `implode(...)` can be replaced with our `$this->createParameterList()` helper.
2. Is the a reason why you can't use `IN` instead of `= ANY`?

**Review by williamjallen — CHANGES_REQUESTED** [review]
Almost ready to go.  A couple more changes and this should be good to merge.  Given that the changes since the last approval are nontrivial, it would be good to get another round of manual testing as well.

File: site/app/libraries/database/DatabaseQueries.php (hunk 18 above)
Comment (dorian451) [reply site/app/libraries/database/DatabaseQueries.php]: 90927284a67635ab386581c2dbbc09175bc4ebb7

File: site/app/libraries/database/DatabaseQueries.php (hunk 17 above)
Comment (dorian451) [reply site/app/libraries/database/DatabaseQueries.php]: Hopefully fixed with fe3c4af9ba9dc551b42772a0c1ad7735e3fe6430

File: site/app/libraries/database/DatabaseQueries.php [hunk 19]
```diff
@@ -2463,6 +2464,52 @@ public function getGradedComponentsCountByGradingSections($g_id, $sections, $sec
...
+            "
+            SELECT 
+                $unit.$section_key, COUNT($unit.$id) as verified_components_count
+            FROM 
+                gradeable_data as gd
+                JOIN gradeable_component_data as gcd ON (gd.gd_id = gcd.gd_id)
+                JOIN $unit ON (gd.gd_$id = $unit.$id)
+            WHERE
+                gd.g_id = ?
+                AND CAST ($unit.$section_key AS TEXT) IN " . $this.createParameterList(count($sections))  . "
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2496]:
```suggestion
                AND CAST ($unit.$section_key AS TEXT) IN " . $this->createParameterList(count($sections)) . "
```

File: site/app/libraries/database/DatabaseQueries.php [hunk 20]
```diff
@@ -2433,6 +2433,7 @@ public function getGradedComponentsCountByGradingSections($g_id, $sections, $sec
             $where = "WHERE active_version > 0 AND ({$section_key} IN " . $this->createParameterList(count($sections)) . ") IS NOT FALSE";
             $params = array_merge($params, $sections);
         }
+
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2436]: This whitespace change should be reverted.

File: site/app/templates/grading/electronic/ta_status/StatusData.twig [hunk 21]
```diff
@@ -61,20 +61,27 @@
...
                 <div style="margin-left: 20px">
+                {# checks if any of sections has a verified value #}
                     {% for key, section in sections %}
                         {% if key != "peer_stu_grad"%}
                                 {% if include_bad_submissions  == 'include'  %}
-                                    {% if key == "NULL" and include_null_section != 'include' %}
-                                        {# Skip rendering this section #}
-                                     {% else %}
-                                        Section {{ key }}: {{ section.graded }} / {{ section.total }} ({{ section.percentage }}%)<br />
+                                    {% if not (key == "NULL" and include_null_section != 'include') %}
```
Comment (williamjallen) [inline site/app/templates/grading/electronic/ta_status/StatusData.twig:68]: It looks like this if-statement is duplicated in both branches of the outer conditional.  Should it be moved outside instead?

**Review by williamjallen — CHANGES_REQUESTED** [review]
It looks like you also have some merge conflicts to address.

File: site/app/templates/grading/electronic/ta_status/StatusData.twig (hunk 21 above)
Comment (dorian451) [reply site/app/templates/grading/electronic/ta_status/StatusData.twig:68]: Fixed in 5c53b4dbd69da2adb776128a462c64d482995bca

File: site/app/libraries/database/DatabaseQueries.php (hunk 19 above)
Comment (dorian451) [reply site/app/libraries/database/DatabaseQueries.php]: fixed in 22d3334b28ff3cde8f82c1bbb520701d377371ef

File: site/app/libraries/database/DatabaseQueries.php (hunk 20 above)
Comment (dorian451) [reply site/app/libraries/database/DatabaseQueries.php:2491]: fixed in 22d3334b28ff3cde8f82c1bbb520701d377371ef

File: site/app/templates/grading/electronic/ta_status/StatusData.twig [hunk 22]
```diff
@@ -61,20 +61,43 @@
...
                 <div style="margin-left: 20px">
+                {# checks if any of sections has a verified value #}
                     {% for key, section in sections %}
                         {% if key != "peer_stu_grad"%}
-                                {% if include_bad_submissions  == 'include'  %}
-                                    {% if key == "NULL" and include_null_section != 'include' %}
-                                        {# Skip rendering this section #}
-                                     {% else %}
-                                        Section {{ key }}: {{ section.graded }} / {{ section.total }} ({{ section.percentage }}%)<br />
+                                {# {% if include_bad_submissions  == 'include'  %}
```
Comment (williamjallen) [inline site/app/templates/grading/electronic/ta_status/StatusData.twig:67]: This commented-out code should be removed.

File: site/app/libraries/database/DatabaseQueries.php [hunk 23]
```diff
@@ -2489,6 +2488,7 @@ public function getGradedComponentsCountByGradingSections($g_id, $sections, $sec
...
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2491]: This whitespace change still needs to be reverted.

File: site/app/libraries/database/DatabaseQueries.php [hunk 24]
```diff
@@ -2432,7 +2432,6 @@ public function getTeamsByGradeableAndRotatingSections($g_id, $sections, $orderB
             $placeholders = implode(",", array_fill(0, count($sections), "?"));
             $params = [$g_id];
             $params = array_merge($params, $sections);
-
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:2435]: Same here.

**Review by williamjallen — CHANGES_REQUESTED** [review]
Once these code cleanliness issues are addressed, I'll approve.

**Review by ziesski — APPROVED** [review]
Beside code clarity mentioned above by @williamjallen, I think the code looks good.

**Review by IDzyre — APPROVED** [review]
Testing changes look good, I haven't tested the functionality.

**Comment by bmcutler** [comment]
@zacharymnp or @ziesski 
Can one of you please manually test and comment that things look good after code changes?

**Comment by ziesski** [comment]
After some stress testing, I noticed some corner case. 
If instructor verifies invalid submission (not enough lateday submission or canceled submission) it gets counted towards existing valid submission. Here is a screenshot example. 
Notice that for section 5 there are 10 students, and 8 of them has valid submissions, and 4 of them has been graded. 
Instructor comes in and verifies just first element of the invalid submission (that mentor graded) and it gets counted as 0.3 out of 8 valid submissions, which may be misleading. 
스크린샷 2024-07-10 오후 2 24 13
스크린샷 2024-07-10 오후 2 24 27


However, I think this is a good stopping point, as it is non-breaking/good change that works with normal use cases, and should get merged, and instead I will create new issue for future fix. @bmcutler
